### PR TITLE
(CDAP-4688) Surface matching artifacts both in SYSTEM and USER scopes…

### PR DESCRIPTION
**Pull request by bdmogal**
… in metadata search, when the target type is either ARTIFACT or ALL.

Jira: [CDAP-4688](https://issues.cask.co/browse/CDAP-4688)
Build: http://builds.cask.co/browse/CDAP-RBT626

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 1]
```diff
@@ -416,11 +419,23 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
...
+   *             {@link MetadataSearchTargetType#ALL} and {@link MetadataSearchTargetType#ARTIFACT}, returns
+   *             matching artifacts in both {@link ArtifactScope#SYSTEM} and {@link ArtifactScope#USER}
    */
-  public List<MetadataEntry> search(final String namespaceId, final String searchQuery,
-                                    final MetadataSearchTargetType type) {
+  public List<MetadataEntry> search(String namespaceId, String searchQuery, MetadataSearchTargetType type) {
+    List<MetadataEntry> searchResults = new ArrayList<>();
+    searchResults.addAll(doSearch(namespaceId, searchQuery, type));
+    // for MetadataSearchTargetType.ARTIFACT and MetadataSearchTargetType.ALL, also search system scope artifacts
+    if (MetadataSearchTargetType.ARTIFACT.equals(type) || MetadataSearchTargetType.ALL.equals(type)) {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: If type is `ALL` then that should automatically search both system and user metadata, right? Why do we need to special case it?

Also, why do we need to special case `ARTIFACT` search?

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: oh this has nothing to do with ``SYSTEM`` and ``USER`` Metadata Scope. This is for ``ArtifactScope``. The issue is that when you are in default namespace for example, and you do a ``list artifacts``, you see both ``ArtifactScope.SYSTEM`` and ``ArtifactScope.USER`` artifacts. However, when you search metadata, without this PR, only ``ArtifactScope.USER`` artifacts are surfaced in the search results.

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: As we discussed, if entities of SYSTEM namespace are considered as part of every namespace, then we should just rewrite all queries to include system scope too.

**Comment by bdmogal** [comment]
@poornachandra addressed comments. Please have another look.

**Comment by bdmogal** [comment]
So this PR is a little different from the description now. There's no special casing for artifacts anymore. In general, all entities from SYSTEM namespace are included in search results - per https://github.com/caskdata/cdap/pull/5076#discussion_r51530495

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 2]
```diff
@@ -466,22 +464,33 @@ private Metadata getSnapshotBeforeTime(Id.NamespacedId targetId, long timeMillis
...
+      String formattedSearchTerm = term.toLowerCase();
+      // if this is a key:value search remove  spaces around the separator too
+      if (formattedSearchTerm.contains(KEYVALUE_SEPARATOR)) {
+        // split the search query in two parts on first occurrence of KEYVALUE_SEPARATOR and the trim the key and value
+        String[] split = formattedSearchTerm.split(KEYVALUE_SEPARATOR, 2);
+        formattedSearchTerm = split[0].trim() + KEYVALUE_SEPARATOR + split[1].trim();
+      }
+      searchTerms.add(namespaceId + KEYVALUE_SEPARATOR + formattedSearchTerm);
+      // also add the system namespace, so entities from system namespace are surfaced in the search results as well
+      searchTerms.add(Id.Namespace.SYSTEM.getId() + KEYVALUE_SEPARATOR + formattedSearchTerm);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: We could add system namespaced search term only if namespace is not equal to system, right?

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 3]
```diff
@@ -404,54 +353,80 @@ public void testSearchOnKeyValue() throws Exception {
...
+
+  @Test
+  public void testSearchIncludesSystemEntities() {
+    // add some metadata to the system artifact
+    Id.Artifact sysArtifact = Id.Artifact.from(
+      Id.Namespace.SYSTEM, new ArtifactId("sysArtifact", new ArtifactVersion("1.0"), ArtifactScope.SYSTEM));
+    String multiWordKey = "multiword";
+    String multiWordValue = "aV1 av2 ,  -  ,  av3 - av4_av5 av6";
+    dataset.setProperty(flow1, multiWordKey, multiWordValue);
+    dataset.setProperty(sysArtifact, multiWordKey, multiWordValue);
```
Comment (poornachandra) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: Would be a good idea to add a non-system artifact, and assert you get both.

**Comment by poornachandra** [comment]
@bdmogal Just two comments, please address them.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 2 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Good point! will do.

**Comment by bdmogal** [comment]
@poornachandra addressed comments

**Comment by rohitsinha54** [comment]
One minor nit. Besides that LGTM

**Comment by bdmogal** [comment]
Build passed. Merging.

**Pull request merged**